boot/bootutil/zephyr/CMakeLists.txt

```diff
@@ -16,6 +16,30 @@ zephyr_library_named(mcuboot_util)
 zephyr_library_sources(
   ../src/bootutil_public.c
     )
+
+if(CONFIG_BOOT_IMAGE_ACCESS_HOOKS)
+  if(NOT CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE STREQUAL "")
+    if(IS_ABSOLUTE ${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+      if(EXISTS ${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+        set(HOOKS_FILE ${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+      endif()
+    elseif((DEFINED CONF_DIR) AND
+           (EXISTS ${CONF_DIR}/${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE}))
+      set(HOOKS_FILE ${CONF_DIR}/${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+    else(EXISTS ${APPLICATION_SOURCE_DIR}/${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+      set(HOOKS_FILE ${APPLICATION_SOURCE_DIR}/${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+    endif()
+  endif()
+
+  if(DEFINED HOOKS_FILE)
+      zephyr_library_sources(
+        ${HOOKS_FILE}
+      )
+  else()
+    message(STATUS "No hooks implementation file.")
+  endif()
+endif()
+
```

Comment:
why aren't we just doing:
```
# <ncs>/<path>/CMakeLists.txt
target_sources(mcuboot_util PRIVATE <hook-file.c>)
```
instead of going through KConfig ?

Reply:
The code was simply moved from other CMakeLists file. Could we leave it as it is for now?
@nvlsianpu, do you have any input here?

Reply:
@tejlmand This path is needed for the upstream. It also allow to add the hook file to the build somewhere else.

Reply:
The script may look complicated but it does its job. I think we can optimize it later, and currently we need it. It does not affect size of binary; just makes build take longer.

Reply:
>  It also allow to add the hook file to the build somewhere else.

Sorry, but i'm not sure I understand what you mean by this ?
What you are requesting is something supported by CMake already, as I showed.

> The code was simply moved from other CMakeLists file. 

Well, it was introduced short time ago, maybe we should do things properly before rolling it out in a release.
https://github.com/mcu-tools/mcuboot/pull/1058

Reply:
Think I found it: https://github.com/nrfconnect/sdk-nrf/blob/master/modules/mcuboot/hooks/nrf53_hooks.c

Reply:
^^ yes this is the file we are using.

Reply:
See alternative proposal which even get rid of the relative / absolute path issue when using Kconfig:
https://github.com/tejlmand/fw-nrfconnect-nrf/commit/b29959dc1f0be337b96403f13c7e4128125f4391
https://github.com/tejlmand/fw-nrfconnect-mcuboot/commit/46c70decc0dd612b2a03ec7a24042fc090f8990e

Reply:
in that alternative we can discuss which CMakeLists.txt file is the most appropriate location for the `target_sources()` code.

Reply:
@tejlmand  what you suggested looks fine to me. Path here is merged in the upstream now. What i want to have upstream is possibility for having multiple build configuration - for having test-build with hooks as well. CMake is not enough to make project configurable (can you propose something else I can use?).
Anyway - User which will use target_source() wil just set `CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE=""`, so this patch will does nothing in that case.

Reply:
> @tejlmand what you suggested looks fine to me. Path here is merged in the upstream now. What i want to have upstream is possibility for having multiple build configuration - for having test-build with hooks as well. CMake is not enough to make project configurable (can you propose something else I can use?).

This is what we do everywhere in Zephyr, we select an option and based on that option CMake determines the file to compile. We never (and should never) point to a c-file from Kconfig.

For example for GPIO, a user can enable GPIO and select a driver:
https://github.com/zephyrproject-rtos/zephyr/blob/e8620a7416e4d51285fe5509e3a5ccbaaaba2818/drivers/gpio/Kconfig.nrfx#L4
```
menuconfig GPIO_NRFX
	bool "nRF GPIO driver"
```
and that results in `gpio_nrfx.c` to be included in the lib.
https://github.com/zephyrproject-rtos/zephyr/blob/e8620a7416e4d51285fe5509e3a5ccbaaaba2818/drivers/gpio/CMakeLists.txt#L21
```
zephyr_library_sources_ifdef(CONFIG_GPIO_NRFX       gpio_nrfx.c)
```

so if mcuboot supports native hook, I would suggest those to be available in same way as we do everywhere else, and not invent a new scheme for that.

Then a downstream user can follow the same principle for adding out-of-tree hooks, like is done for entropy.
https://github.com/nrfconnect/sdk-nrf/blob/f10756d98c5bec203971771d8700ad81aa0283d4/drivers/entropy/Kconfig#L7-L8
```
config ENTROPY_CC3XX
	bool "Arm CC3XX RNG driver for Nordic devices" if !BUILD_WITH_TFM
```
which results in:
https://github.com/nrfconnect/sdk-nrf/blob/f10756d98c5bec203971771d8700ad81aa0283d4/drivers/entropy/CMakeLists.txt#L6-L7
```
zephyr_library_amend()
zephyr_library_sources_ifdef(CONFIG_ENTROPY_CC3XX entropy_cc310.c)
```
or they can resort to plain cmake if they like.

> Anyway - User which will use target_source() wil just set CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE="", so this patch will does nothing in that case.

Besides not following the pattern we use everywhere else, then it will also result in this message without any guidance on what might be wrong:
```
  else()
    message(STATUS "No hooks implementation file.")
  endif()
```

Reply:
I think you are right @tejlmand. Will register this as issue upstream an will consult with you regard the solution when will I will be fixing that. As far I understand I can just remove whole complex patch resolution from the CMake and add proper hint to the Kconfig help. For doing upstream test-build i can add hooks_saple.c in upstream camake depending on a build commanding option.

Reply:
as soon as such issue has been created, post a link and I'll approve this PR, as I don't believe I have any other blocking comments.

Reply:
Marek removed this commit as flash_nop_driver was enough to mitigate the issue which hooks amendment tried to fix.